File: ios/KakaoSdkUser/UserApiClientModule.swift
//
//  UserApiClientModule.swift
//  react-native-kakao-sdk-all
//
//  Created by kakao-tam on 2022/05/06.
//

import Foundation

import KakaoSDKAuth
import KakaoSDKUser

@objc(UserApiClientModule)
class UserApiClientModule: NSObject {
    
    @objc(accessTokenInfo:withReject:)
    func accessTokenInfo(_ resolve:@escaping RCTPromiseResolveBlock,
                         reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            UserApi.shared.accessTokenInfo {
                (accessTokenInfo : AccessTokenInfo?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(accessTokenInfo?.toJsonString());
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(certLoginWithKakaoAccount:withResolve:withReject:)
    func certLoginWithKakaoAccount(_ param:NSDictionary,
                                   resolve:@escaping RCTPromiseResolveBlock,
                                   reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(LoginParam.self)!
            UserApi.shared.certLoginWithKakaoAccount(prompts: p.prompts,
                                                     state: p.state,
                                                     loginHint: p.loginHint,
                                                     nonce: p.nonce) {
                (certTokenInfo : CertTokenInfo?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(certTokenInfo?.toJsonString());
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(certLoginWithKakaoTalk:withResolve:withReject:)
    func certLoginWithKakaoTalk(_ param:NSDictionary,
                                resolve:@escaping RCTPromiseResolveBlock,
                                reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(LoginParam.self)!
            UserApi.shared.certLoginWithKakaoTalk(prompts: p.prompts,
                                                  state: p.state,
                                                  channelPublicIds: p.channelPublicIds,
                                                  serviceTerms: p.serviceTerms,
                                                  nonce: p.nonce) {
                (certTokenInfo : CertTokenInfo?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(certTokenInfo?.toJsonString());
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(isKakaoTalkLoginAvailable:withReject:)
    func isKakaoTalkLoginAvailable(_ resolve:@escaping RCTPromiseResolveBlock,
                                   reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            resolve(UserApi.isKakaoTalkLoginAvailable())
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(loginWithKakaoAccount:withResolve:withReject:)
    func loginWithKakaoAccount(_ param:NSDictionary,
                               resolve:@escaping RCTPromiseResolveBlock,
                               reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(LoginParam.self)!
            if let scopes = p.scopes {
                UserApi.shared.loginWithKakaoAccount(scopes: scopes,
                                                     nonce: p.nonce) {
                    (oauthToken: OAuthToken?, error: Error?) in
                    if let error = error {
                        reject(error.rnCode(), error.rnMessage(), error)
                    } else {
                        resolve(oauthToken?.toJsonString())
                    }
                }
            } else {
                if p.channelPublicIds != nil || p.serviceTerms != nil {
                    UserApi.shared.loginWithKakaoAccount(prompts: p.prompts,
                                                         channelPublicIds: p.channelPublicIds,
                                                         serviceTerms: p.serviceTerms,
                                                         nonce: p.nonce) {
                        (oauthToken: OAuthToken?, error: Error?) in
                        if let error = error {
                            reject(error.rnCode(), error.rnMessage(), error)
                        } else {
                            resolve(oauthToken?.toJsonString())
                        }
                    }
                } else {
                    UserApi.shared.loginWithKakaoAccount(prompts: p.prompts,
                                                         loginHint: p.loginHint,
                                                         nonce: p.nonce) {
                        (oauthToken: OAuthToken?, error: Error?) in
                        if let error = error {
                            reject(error.rnCode(), error.rnMessage(), error)
                        } else {
                            resolve(oauthToken?.toJsonString())
                        }
                    }
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(loginWithKakaoTalk:withResolve:withReject:)
    func loginWithKakaoTalk(_ param:NSDictionary,
                            resolve:@escaping RCTPromiseResolveBlock,
                            reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(LoginParam.self)!
            UserApi.shared.loginWithKakaoTalk(channelPublicIds: p.channelPublicIds,
                                              serviceTerms: p.serviceTerms,
                                              nonce: p.nonce) {
                (oauthToken: OAuthToken?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(oauthToken?.toJsonString())
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }

    @objc(logout:withReject:)
    func logout(_ resolve:@escaping RCTPromiseResolveBlock,
                reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            UserApi.shared.logout() {
                (error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(true)
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(me:withResolve:withReject:)
    func me(_ param:NSDictionary,
            resolve:@escaping RCTPromiseResolveBlock,
            reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(MeParam.self)!
            UserApi.shared.me(propertyKeys: p.propertyKeys,
                              secureResource: p.secureResource) {
                (user: User?, error:Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(user?.toJsonString())
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(revokeScopes:withResolve:withReject:)
    func revokeScopes(_ param:NSDictionary,
                      resolve:@escaping RCTPromiseResolveBlock,
                      reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(ScopesParam.self)!
            UserApi.shared.revokeScopes(scopes: p.scopes ?? []){
                (scopeInfo: ScopeInfo?, error:Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(scopeInfo?.toJsonString())
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(scopes:withResolve:withReject:)
    func scopes(_ param:NSDictionary,
                resolve:@escaping RCTPromiseResolveBlock,
                reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(ScopesParam.self)!
            UserApi.shared.scopes(scopes: p.scopes ?? []) {
                (scopeInfo: ScopeInfo?, error:Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(scopeInfo?.toJsonString())
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(serviceTerms:withResolve:withReject:)
    func serviceTerms(_ param:NSDictionary,
                      resolve:@escaping RCTPromiseResolveBlock,
                      reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(ServiceTermsParam.self)!
            UserApi.shared.serviceTerms(extra: p.extra) {
                (userServiceTerms: UserServiceTerms?, error:Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(userServiceTerms?.toJsonString())
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(shippingAddresses:withResolve:withReject:)
    func shippingAddresses(_ param:NSDictionary,
                           resolve:@escaping RCTPromiseResolveBlock,
                           reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(ShippingAddressesParam.self)!
            if p.addressId != nil {
                UserApi.shared.shippingAddresses(addressId: p.addressId!) {
                    (userShippingAddresses: UserShippingAddresses?, error:Error?) in
                    if let error = error {
                        reject(error.rnCode(), error.rnMessage(), error)
                    } else {
                        resolve(userShippingAddresses?.toJsonString())
                    }
                }
            } else {
                UserApi.shared.shippingAddresses(fromUpdatedAt: p.fromUpdatedAt,
                                                 pageSize: p.pageSize) {
                    (userShippingAddresses: UserShippingAddresses?, error:Error?) in
                    if let error = error {
                        reject(error.rnCode(), error.rnMessage(), error)
                    } else {
                        resolve(userShippingAddresses?.toJsonString())
                    }
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(signup:withResolve:withReject:)
    func signup(_ param:NSDictionary,
                resolve:@escaping RCTPromiseResolveBlock,
                reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(SignupParam.self)!
            UserApi.shared.signup(properties: p.properties) {
                (id: Int64?, error:Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(id)
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(unlink:withReject:)
    func unlink(_ resolve:@escaping RCTPromiseResolveBlock,
                reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            UserApi.shared.unlink() {
                (error:Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(true)
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(updateProfile:withResolve:withReject:)
    func updateProfile(_ param:NSDictionary,
                       resolve:@escaping RCTPromiseResolveBlock,
                       reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(UpdateProfileParam.self)!
            UserApi.shared.updateProfile(properties: p.properties ?? [:]) {
                (error:Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(true)
                }
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
}

File: android/src/main/java/com/kakaotam/rnsdk/user/UserApiClientModule.kt
package com.kakaotam.rnsdk.user

import com.facebook.react.bridge.*
import com.kakao.sdk.auth.AuthCodeClient
import com.kakao.sdk.auth.model.CertTokenInfo
import com.kakao.sdk.auth.model.OAuthToken
import com.kakao.sdk.user.UserApiClient
import com.kakao.sdk.user.model.*
import com.kakaotam.rnsdk.common.PromiseTransformer
import com.kakaotam.rnsdk.common.ReactToParam
import com.kakaotam.rnsdk.common.SdkInitializer
import com.kakaotam.rnsdk.user.model.*

class UserApiClientModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String {
        return "UserApiClientModule"
    }

    @ReactMethod
    private fun accessTokenInfo(promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        UserApiClient.instance.accessTokenInfo f@{ accessTokenInfo: AccessTokenInfo?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(accessTokenInfo))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun certLoginWithKakaoAccount(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, LoginParam::class.java)!!
        UserApiClient.instance.certLoginWithKakaoAccount(
            reactContext.currentActivity!!,
            p.prompts,
            p.state,
            p.nonce,
            p.channelPublicIds,
            p.serviceTerms,
            p.loginHint
        ) f@{ certTokenInfo: CertTokenInfo?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(certTokenInfo))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun certLoginWithKakaoTalk(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, LoginParam::class.java)!!
        UserApiClient.instance.certLoginWithKakaoTalk(
            reactContext.currentActivity!!,
            p.prompts,
            p.state,
            AuthCodeClient.DEFAULT_REQUEST_CODE,
            p.nonce,
            p.channelPublicIds,
            p.serviceTerms
        ) f@{ certTokenInfo: CertTokenInfo?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(certTokenInfo))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun isKakaoTalkLoginAvailable(promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        promise.resolve(UserApiClient.instance.isKakaoTalkLoginAvailable(reactContext.currentActivity!!))
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun loginWithKakaoAccount(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, LoginParam::class.java)!!
        p.scopes?.let { scopes ->
            UserApiClient.instance.loginWithNewScopes(
                reactContext.currentActivity!!,
                scopes,
                p.nonce
            ) f@{ oAuthToken: OAuthToken?, error: Throwable? ->
                error?.let {
                    promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                    return@f
                }
                promise.resolve(PromiseTransformer.toJsonString(oAuthToken))
            }
        } ?: run {
            UserApiClient.instance.loginWithKakaoAccount(
                reactContext.currentActivity!!,
                p.prompts,
                p.loginHint,
                p.nonce,
                p.channelPublicIds,
                p.serviceTerms
            ) f@{ oAuthToken: OAuthToken?, error: Throwable? ->
                error?.let {
                    promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                    return@f
                }
                promise.resolve(PromiseTransformer.toJsonString(oAuthToken))
            }
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun loginWithKakaoTalk(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, LoginParam::class.java)!!
        UserApiClient.instance.loginWithKakaoTalk(
            reactContext.currentActivity!!,
            AuthCodeClient.DEFAULT_REQUEST_CODE,
            p.nonce,
            p.channelPublicIds,
            p.serviceTerms
        ) f@{ oAuthToken: OAuthToken?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(oAuthToken))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

//    @ReactMethod
//    private fun loginWithNewScopes(param: ReadableMap, promise: Promise) = try {
//        SdkInitializer.ensure(reactContext)
//        val p = ReactToParam.toParam(param, LoginWithNewScopesParam::class.java)!!
//        UserApiClient.instance.loginWithNewScopes(
//            reactContext.currentActivity!!,
//            p.scopes,
//            p.nonce
//        ) f@{ oAuthToken: OAuthToken?, error: Throwable? ->
//            error?.let {
//                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
//                return@f
//            }
//            promise.resolve(PromiseTransformer.toJsonString(oAuthToken))
//        }
//    } catch (e: Exception) {
//        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
//    }

    @ReactMethod
    private fun logout(promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        UserApiClient.instance.logout f@{ error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(true)
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun me(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, MeParam::class.java)!!
        UserApiClient.instance.me(p.secureReSource) f@{ user: User?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(user))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun revokeScopes(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, ScopesParam::class.java)!!
        UserApiClient.instance.revokeScopes(p.scopes) f@{ scopeInfo: ScopeInfo?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(scopeInfo))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun scopes(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, ScopesParam::class.java)!!
        UserApiClient.instance.scopes(p.scopes) f@{ scopeInfo: ScopeInfo?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(scopeInfo))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun serviceTerms(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, ServiceTermsParam::class.java)!!
        UserApiClient.instance.serviceTerms(p.extra) f@{ userServiceTerms: UserServiceTerms?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(userServiceTerms))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun shippingAddresses(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, ShippingAddressesParam::class.java)!!
        p.addressId?.let {
            UserApiClient.instance.shippingAddresses(it) f@{ userShippingAddresses: UserShippingAddresses?, error: Throwable? ->
                error?.let {
                    promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                    return@f
                }
                promise.resolve(PromiseTransformer.toJsonString(userShippingAddresses))
                return@f
            }
        } ?: run {
            UserApiClient.instance.shippingAddresses(
                p.fromUpdateAt,
                p.pageSize
            ) f@{ userShippingAddresses: UserShippingAddresses?, error: Throwable? ->
                error?.let {
                    promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                    return@f
                }
                promise.resolve(PromiseTransformer.toJsonString(userShippingAddresses))
                return@f
            }
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun signup(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, SignupParam::class.java)!!
        UserApiClient.instance.signup(p.properties) f@{ error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(true)
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun unlink(promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        UserApiClient.instance.unlink() f@{ error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(true)
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun updateProfile(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, UpdateProfileParam::class.java)!!
        UserApiClient.instance.updateProfile(p.properties) f@{ error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(true)
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }
}
